File: src/devtools/executor.py
from __future__ import annotations

import contextlib
import importlib.util
import io
import uuid
from pathlib import Path
from typing import Callable

from devtools.models import AppConfig, ExecutionResult, Project, ScriptContext, ScriptDefinition


def run_script_for_projects(
    config: AppConfig,
    script: ScriptDefinition,
    projects: list[Project],
    cli_args: dict[str, object] | None = None,
    event_callback: Callable[[str], None] | None = None,
) -> list[ExecutionResult]:
    results: list[ExecutionResult] = []
    for project in projects:
        if event_callback:
            event_callback(f"[start] {script.script_id} -> {project.path}")
        result = run_script_for_project(config, script, project, cli_args=cli_args)
        results.append(result)
        if event_callback:
            prefix = "ok" if result.success else "fail"
            detail = result.message if result.message else result.error
            event_callback(f"[{prefix}] {project.path} :: {detail}")
    return results


def run_script_for_project(
    config: AppConfig,
    script: ScriptDefinition,
    project: Project,
    cli_args: dict[str, object] | None = None,
) -> ExecutionResult:
    args = dict(script.default_args)
    if cli_args:
        args.update(cli_args)

    context = ScriptContext(
        config_path=config.config_path,
        script=script,
        project=project,
        args=args,
        run_id=uuid.uuid4().hex,
    )
    stdout = io.StringIO()
    stderr = io.StringIO()
    try:
        runner = _load_entry(script)
        with contextlib.redirect_stdout(stdout), contextlib.redirect_stderr(stderr):
            response = runner(context)
        success = bool(response.get("success", True)) if isinstance(response, dict) else True
        message = str(response.get("message", "")) if isinstance(response, dict) else ""
        return ExecutionResult(
            project=project,
            script=script,
            success=success,
            message=message,
            output=stdout.getvalue(),
            error=stderr.getvalue(),
        )
    except Exception as exc:  # noqa: BLE001
        return ExecutionResult(
            project=project,
            script=script,
            success=False,
            message="",
            output=stdout.getvalue(),
            error=f"{stderr.getvalue()}{exc}",
        )


def _load_entry(script: ScriptDefinition):
    module_rel, _, function_name = script.entry.partition(":")
    if not module_rel or not function_name:
        raise ValueError(f"Invalid entry definition for {script.script_id}: {script.entry}")

    module_path = (script.directory / module_rel).resolve()
    if not module_path.exists():
        raise FileNotFoundError(f"Entry module not found: {module_path}")

    module_name = f"devtools_script_{script.script_id}"
    spec = importlib.util.spec_from_file_location(module_name, module_path)
    if spec is None or spec.loader is None:
        raise ImportError(f"Could not load module from {module_path}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    entry = getattr(module, function_name, None)
    if entry is None:
        raise AttributeError(f"Function {function_name} not found in {module_path}")
    return entry

File: src/executor.ts
import { randomUUID } from "node:crypto";
import path from "node:path";
import { pathToFileURL } from "node:url";

import type { AppConfig, ExecutionResult, Project, RunOutputMode, ScriptContext, ScriptDefinition } from "./models.ts";

export async function runScriptForProjects(
  config: AppConfig,
  script: ScriptDefinition,
  projects: Project[],
  cliArgs: Record<string, unknown> = {},
  eventCallback?: (message: string) => void,
  signal?: AbortSignal,
  outputMode: RunOutputMode = "capture",
): Promise<ExecutionResult[]> {
  const batchRunId = randomUUID().replaceAll("-", "");
  if (script.scope === "global") {
    eventCallback?.(`[start] ${script.scriptId} -> global`);
    const result = await runScriptForProject(config, script, undefined, cliArgs, eventCallback, signal, outputMode, {
      batchRunId,
      batchProjectIndex: 0,
      batchProjectCount: 0,
    });
    const prefix = result.success ? "ok" : "fail";
    const detail = result.message || result.error;
    eventCallback?.(`[${prefix}] global :: ${detail}`);
    return [result];
  }

  const results: ExecutionResult[] = [];
  for (const [projectIndex, project] of projects.entries()) {
    if (signal?.aborted) {
      eventCallback?.("[cancelled] Script execution interrupted.");
      break;
    }
    if (!supportsScript(script, project)) {
      const result: ExecutionResult = {
        project,
        script,
        success: true,
        message: `Skipped: ${script.scriptId} does not apply to [${project.projectTypes.join(",")}].`,
        output: "",
        error: "",
      };
      results.push(result);
      eventCallback?.(`[skip] ${project.path} :: ${result.message}`);
      continue;
    }
    writeProjectBanner(project, outputMode);
    eventCallback?.(`[start] ${script.scriptId} -> ${project.path}`);
    const result = await runScriptForProject(config, script, project, cliArgs, eventCallback, signal, outputMode, {
      batchRunId,
      batchProjectIndex: projectIndex,
      batchProjectCount: projects.length,
    });
    results.push(result);
    const prefix = result.success ? "ok" : "fail";
    const detail = result.message || result.error;
    eventCallback?.(`[${prefix}] ${project.path} :: ${detail}`);
  }
  return results;
}

function writeProjectBanner(project: Project, outputMode: RunOutputMode): void {
  if (outputMode !== "passthrough") {
    return;
  }
  const separator = "-".repeat(Math.max(24, project.name.length + 16));
  process.stdout.write(
    [
      "",
      `\x1b[93m${separator}\x1b[0m`,
      `\x1b[30;103m----- ${project.name} -----\x1b[0m`,
      `\x1b[93m${separator}\x1b[0m`,
      "",
    ].join("\n"),
  );
}

export async function runScriptForProject(
  config: AppConfig,
  script: ScriptDefinition,
  project?: Project,
  cliArgs: Record<string, unknown> = {},
  eventCallback?: (message: string) => void,
  signal?: AbortSignal,
  outputMode: RunOutputMode = "capture",
  batchMetadata?: {
    batchRunId: string;
    batchProjectIndex: number;
    batchProjectCount: number;
  },
): Promise<ExecutionResult> {
  const args = { ...script.defaultArgs, ...cliArgs };
  const context: ScriptContext = {
    configPath: config.configPath,
    script,
    project,
    args,
    runId: randomUUID().replaceAll("-", ""),
    batchRunId: batchMetadata?.batchRunId,
    batchProjectIndex: batchMetadata?.batchProjectIndex,
    batchProjectCount: batchMetadata?.batchProjectCount,
    log: eventCallback,
    signal,
    outputMode,
  };

  let stdoutBuffer = "";
  let stderrBuffer = "";
  const shouldCaptureOutput = outputMode === "capture";
  const originalStdoutWrite = process.stdout.write.bind(process.stdout);
  const originalStderrWrite = process.stderr.write.bind(process.stderr);

  const interceptStdout = (chunk: string | Uint8Array): boolean => {
    stdoutBuffer += typeof chunk === "string" ? chunk : Buffer.from(chunk).toString("utf8");
    return true;
  };
  const interceptStderr = (chunk: string | Uint8Array): boolean => {
    stderrBuffer += typeof chunk === "string" ? chunk : Buffer.from(chunk).toString("utf8");
    return true;
  };

  try {
    if (signal?.aborted) {
      return {
        project,
        script,
        success: false,
        message: "Execution cancelled before start.",
        output: "",
        error: "",
      };
    }
    const runner = await loadEntry(script);
    if (shouldCaptureOutput) {
      process.stdout.write = interceptStdout as typeof process.stdout.write;
      process.stderr.write = interceptStderr as typeof process.stderr.write;
    }
    const response = await runner(context);
    return {
      project,
      script,
      success: Boolean(response.success),
      message: String(response.message ?? ""),
      output: stdoutBuffer,
      error: stderrBuffer,
    };
  } catch (error) {
    return {
      project,
      script,
      success: false,
      message: "",
      output: stdoutBuffer,
      error: `${stderrBuffer}${error instanceof Error ? error.message : String(error)}`,
    };
  } finally {
    if (shouldCaptureOutput) {
      process.stdout.write = originalStdoutWrite;
      process.stderr.write = originalStderrWrite;
    }
  }
}

async function loadEntry(script: ScriptDefinition) {
  const modulePath = path.join(script.directory, "script.ts");
  const scriptModule = await import(pathToFileURL(modulePath).href);
  const runner = scriptModule[script.entry];
  if (!runner) {
    throw new Error(`Unknown script entry for ${script.scriptId}: ${script.entry} in ${modulePath}`);
  }
  return runner;
}

function supportsScript(script: ScriptDefinition, project: Project): boolean {
  if (script.scope === "global") {
    return true;
  }
  return script.projectTypes.some((projectType) => project.projectTypes.includes(projectType));
}
